pkg/check/node_perf.go

```diff
@@ -0,0 +1,207 @@
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+)
+
+// CheckNodeDiskPerf Checks node performance of master nodes.
+type CheckNodeDiskPerf struct{}
+
+func (c *CheckNodeDiskPerf) Name() string {
+	return "CheckNodePerf"
+}
+
+type MetricCheckDef struct {
+	warningLevel        int64
+	errorLevel          int64
+	metricName          string
+	warningEventMessage string
+	errorEventMessage   string
+	checkMasters        bool
+	checkWorkers        bool
+}
+
+const (
+	metricMaxSamples = 300
+	metricIntervalID = 20
+)
+
+var (
+	perfMetricIDMap       = map[string]*types.PerfMetricId{}
+	buildCounterIDMapLock sync.Mutex
+
+	metricCheckDefs = []MetricCheckDef{
+		{
+			warningLevel:        100,
+			errorLevel:          1000,
+			metricName:          "disk.maxTotalLatency.latest",
+			warningEventMessage: "the node has disk latency of greater than 100ms",
+			errorEventMessage:   "the node has disk latency of greater than 1000ms",
+			checkMasters:        true,
+		},
+	}
+)
+
+func (c *CheckNodeDiskPerf) BuildCounterIdMap(ctx *CheckContext, vm *mo.VirtualMachine) error {
+	buildCounterIDMapLock.Lock()
+	defer buildCounterIDMapLock.Unlock()
+
+	if len(perfMetricIDMap) > 0 {
+		return nil
+	}
+	var counterIds []int32
+
+	perf := types.QueryAvailablePerfMetric{
+		This:   *ctx.VMClient.ServiceContent.PerfManager,
+		Entity: vm.ManagedEntity.Reference(),
+	}
+	tctx, cancel := context.WithTimeout(ctx.Context, *Timeout)
+	defer cancel()
+
+	perfList, err := methods.QueryAvailablePerfMetric(tctx, ctx.VMClient.RoundTripper, &perf)
+	if err != nil {
+		klog.V(2).Infof("error getting available performance metrics: %s", err.Error())
+		return err
+	}
+
+	var counterIDPerfMetricIDMap = map[int32]*types.PerfMetricId{}
+	klog.V(4).Infof("found [%d] performance metrics available", len(perfList.Returnval))
+	for _, perf := range perfList.Returnval {
+		counterIds = append(counterIds, perf.CounterId)
+		copyPerf := perf
+		counterIDPerfMetricIDMap[perf.CounterId] = &copyPerf
+	}
+
+	perfCounter := types.QueryPerfCounter{
+		This:      *ctx.VMClient.ServiceContent.PerfManager,
+		CounterId: counterIds,
+	}
+
+	tctx, cancel = context.WithTimeout(ctx.Context, *Timeout)
+	defer cancel()
+
+	perfCounters, err := methods.QueryPerfCounter(tctx, ctx.VMClient.RoundTripper, &perfCounter)
+	if err != nil {
+		return err
+	}
+
+	for _, perfCounter := range perfCounters.Returnval {
+		if _, ok := counterIDPerfMetricIDMap[perfCounter.Key]; !ok {
+			continue
+		}
+		perfMetricIDMap[perfCounter.Name()] = counterIDPerfMetricIDMap[perfCounter.Key]
+	}
+	return nil
+}
+
+func (c *CheckNodeDiskPerf) GetPerfMetric(ctx *CheckContext, vm *mo.VirtualMachine, metricName string) (*types.PerfMetricId, error) {
+	err := c.BuildCounterIdMap(ctx, vm)
+	if err != nil {
+		return nil, err
+	}
+	if _, ok := perfMetricIDMap[metricName]; !ok {
+		return nil, fmt.Errorf("unable to find metric with name[%s]", metricName)
```

Comment:
This is failing in CI environment with:

```
I0217 15:39:05.871257       1 operator.go:271] CheckNodePerf:ci-op-wfwps240-d1161-gzs57-master-0 failed: unable to find metric with name[disk.maxTotalLatency.latest]
I0217 15:39:05.872328       1 operator.go:271] CheckNodePerf:ci-op-wfwps240-d1161-gzs57-master-1 failed: unable to find metric with name[disk.maxTotalLatency.latest]
I0217 15:39:05.872773       1 operator.go:271] CheckNodePerf:ci-op-wfwps240-d1161-gzs57-master-2 failed: unable to find metric with name[disk.maxTotalLatency.latest]
I0217 15:47:08.406262       1 operator.go:271] CheckNodePerf:ci-op-wfwps240-d1161-gzs57-master-2 failed: unable to find metric with name[disk.maxTotalLatency.latest]
```

Do we want these to be a hard failure?

Reply:
There is something weird going on and I think nodes are being listed twice?

```
0217 15:35:03.794543       1 operator.go:274] CheckNodePerf:ci-op-wfwps240-d1161-gzs57-worker-wssxh passed
I0217 15:39:05.867338       1 operator.go:274] CheckNodePerf:ci-op-wfwps240-d1161-gzs57-worker-wssxh passed
I0217 15:39:05.868738       1 operator.go:274] CheckNodePerf:ci-op-wfwps240-d1161-gzs57-worker-gvdl5 passed
I0217 15:39:05.870926       1 operator.go:274] CheckNodePerf:ci-op-wfwps240-d1161-gzs57-worker-cz9wq passed
I0217 15:39:05.871257       1 operator.go:271] CheckNodePerf:ci-op-wfwps240-d1161-gzs57-master-0 failed: unable to find metric with name[disk.maxTotalLatency.latest]
I0217 15:39:05.872328       1 operator.go:271] CheckNodePerf:ci-op-wfwps240-d1161-gzs57-master-1 failed: unable to find metric with name[disk.maxTotalLatency.latest]
I0217 15:39:05.872773       1 operator.go:271] CheckNodePerf:ci-op-wfwps240-d1161-gzs57-master-2 failed: unable to find metric with name[disk.maxTotalLatency.latest]
I0217 15:47:08.406262       1 operator.go:271] CheckNodePerf:ci-op-wfwps240-d1161-gzs57-master-2 failed: unable to find metric with name[disk.maxTotalLatency.latest]
```

Reply:
I'm checking this out